protocol/opcm-v2.md

```diff
@@ -0,0 +1,234 @@
+- **Maintainability**: Unified code paths for deploy and upgrade operations
+- **Developer experience**: Minimize knowledge burden through auto-generated glue code
+- **Release readiness**: Always in deployable state without post-upgrade cleanup
+
+### Constraints
+
+- Must integrate with existing op-deployer Go tooling
+- Cannot reduce any existing functionality from OPCM v1
+- Interface changes are acceptable but must serve same purposes
+- Minor breaking changes acceptable only if well-justified and documented
+- Must handle existing deployed chains and future chains uniformly
+
+## Proposed Solution
+
+OPContractsManager v2 redesigns the architecture around three core principles: **interface simplification**, **unified initialization**, and **generated integration code**.
+
+### Architecture Overview
+
+OPCM v2 consolidates the 5 separate functions of v1 into 2 primary functions:
+- `deploy(Config memory config)`: Deploys a new OP Stack chain
+- `upgrade(Config memory config)`: Upgrades an existing chain
+
+Both functions share the same initialization path for contracts, eliminating the special upgrade functions that previously required maintenance between releases.
+
+### Key Architectural Changes
+
+**1. Dynamic Dispute Game Interface**
+
+Replace hardcoded assumptions about game types with dynamic arrays:
+```solidity
+struct GameConfig {
+    GameType gameType;
+    address implementation;
+    bytes32 prestate;
+}
+
+struct Config {
+    // ... other config
+    GameConfig[] games;  // Support arbitrary number of game types
+}
+```
+
+This eliminates the need for `addGameType()` and `updatePrestate()` as separate functions—these operations become implicit in `upgrade()` when the games array differs from the current state.
+
+**2. Unified Initialization Pattern**
+
+Both `deploy()` and `upgrade()` use the same contract initialization path:
+- Clear the `initialized` slot (for upgrades)
+- Call the standard `initialize()` function
+- No special upgrade functions that need removal in subsequent releases
+
+This ensures:
+- Single code path for both operations
+- Always release-ready (no post-upgrade cleanup)
+- Easier testing and maintenance
+
+**3. Auto-Generated Integration Code**
+
+Generate the Go glue code between OPCM and op-deployer from Solidity interfaces:
+- Solidity function signatures/ABIs are the source of truth
+- Go code for constructing function calls is automatically generated
+- Developers only need to add new CLI arguments to op-deployer when introducing net-new inputs
+- Reduces knowledge burden—developers don't need to understand integration layer
+
+### Implementation Flow
+
+Both `deploy()` and `upgrade()` follow a unified three-phase flow:
+
+**Phase 1: Chain World Assembly (Load-or-Deploy Pattern)**
+
+Both functions begin by gathering the complete set of contracts for the chain—the "chain world"—which includes SystemConfig, OptimismPortal, AnchorStateRegistry, and all other chain contracts.
```

Comment:
Does this include the Superchain contracts? How do we upgrade those?

Reply:
good callout. The doc mentions 5 functions  (`deploy()`, `upgrade()`, `addGameType()`, `updatePrestate()`, `interopMigrate()`), but there is now also `upgradeSuperchainConfig()`.